### ISSUE TITLE
Failed upload causes the next request to timeout

### ISSUE THREAD

**Issue by peterbandi** [fafhrd91/aiohttp#7599]
### Describe the bug

If a request with POST method using `aiohttp` gets a _400 Bad Request_ response the next request either times out raising an `asyncio.exceptions.TimeoutError` or gets a _400 Bad Request_ response.

The same requests implemented the blocking `requests` library does not have this issue.

### To Reproduce

**Python environment (conda)**
```
channels:
  - conda-forge
  - defaults
dependencies:
  - aiohttp=3.8.5=py39h5eee18b_0
  - python=3.9.17=h955ad1f_0
  - requests=2.31.0=pyhd8ed1ab_0
```

**Server: server_aiohttp.py**
```python 
import aiohttp.web

routes = aiohttp.web.RouteTableDef()

@routes.get(path="/check")
async def path_check(request):
    return aiohttp.web.Response(text='{"message": "Check!"}')


@routes.post(path="/wrong")
async def path_wrong(request):
    return aiohttp.web.Response(status=400, text='{"message": "Wrong?"}')

def init(argv):
    app = aiohttp.web.Application()
    app.add_routes(routes)
    return app
```

**Reference client with Requests: client_requests.py**
```python
import tempfile
from pathlib import Path
import requests

def main():
    url = "http://localhost:8000"
    check_url = f"{url}/check"  # Test URL to see if it works.
    upload_url = f"{url}/wrong"  # URL to use for upload.
    file_size_bytes = 1024 * 1024  # File size to upload.

    with tempfile.TemporaryDirectory() as temp_dir_name:
        # Create a temporary file to send through.
        file_path = Path(temp_dir_name, "file")
        with open(file=file_path, mode="w") as file:
            file.write("0" * file_size_bytes)

        with requests.Session() as session:
            # Not necessary, just to test if the connection works.
            response = session.get(url=check_url)
            print(f"{check_url} [{response.status_code}] {response.text}")

            # Try to upload a file. The issue is the same with non-existing path like "/nope".
            with open(file=file_path, mode="rb") as file:
                response = session.post(url=upload_url, files={"file": file})
                print(f"{upload_url} [{response.status_code}] {response.text}")

            # Check if it still works.
            response = session.get(url=check_url)
            print(f"{check_url} [{response.status_code}] {response.text}")

if __name__ == "__main__":
    main()
```

**Client with AIOHTTP: client_aiohttp.py**
```python
import asyncio
import tempfile
from pathlib import Path
import aiohttp

async def main():
    url = "http://localhost:8000"
    check_url = f"{url}/check"  # Test URL to see if it works.
    upload_url = f"{url}/wrong"  # URL to use for upload.
    file_size_bytes = 1024 * 1024  # File size to upload.

    with tempfile.TemporaryDirectory() as temp_dir_name:
        # Create a temporary file to send through.
        file_path = Path(temp_dir_name, "file")
        with open(file=file_path, mode="w") as file:
            file.write("0" * file_size_bytes)

        async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=5)) as session:
            # Not necessary, just to test if the connection works.
            response = await session.request(method="GET", url=check_url)
            response_content = await response.content.read()
            print(f"{check_url} [{response.status}] {response_content.decode()}")

            # Try to upload a file. The issue is the same with non-existing path like "/nope".
            with open(file=file_path, mode="rb") as file:
                response = await session.request(method="POST", url=upload_url, data={"file": file})
                response_content = await response.content.read()
                print(f"{upload_url} [{response.status}] {response_content.decode()}")

            # Check if it still works.
            # - 1M upload file size: this request always times out.
            # - 128K upload file size: this request tries out. On subsequent run attempts it returns 400.
            # - 1K upload file size: this request always returns 400.
            response = await session.request(method="GET", url=check_url)
            response_content = await response.content.read()
            print(f"{check_url} [{response.status}] {response_content.decode()}")

if __name__ == "__main__":
    asyncio.run(main=main())
```

**Steps to reproduce**
1. Start the server with: `python -m aiohttp.web -H localhost -P 8000 server_aiohttp:init` command
2. Execute the _client_requests.py_ reference script, the output should be:
```
http://localhost:8000/check [200] {"message":"Check!"}
http://localhost:8000/wrong [400] {"message":"Wrong?"}
http://localhost:8000/check [200] {"message":"Check!"}
```
3. Execute the _client_aiohttp.py_ script, observe the raised exception:
```
http://localhost:8000/check [200] {"message":"Check!"}
http://localhost:8000/wrong [400] {"message":"Wrong?"}
...
asyncio.exceptions.TimeoutError
```
4. Observe the console of the server. It doesn't log the 3rd request from `aiohttp`.

**Steps to reproduce with smaller file**
1. Start the server with: `python -m aiohttp.web -H localhost -P 8000 server_aiohttp:init` command
2. Modify the variable to `file_size_bytes = 1024` to reduce the size of the uploaded file to 1K.
3. Execute the _client_requests.py_ reference script, the output should be:
```
http://localhost:8000/check [200] {"message":"Check!"}
http://localhost:8000/wrong [400] {"message":"Wrong?"}
http://localhost:8000/check [200] {"message":"Check!"}
```
4. Execute the _client_aiohttp.py_ script, observe the raised exception:
```
http://localhost:8000/check [200] {"message":"Check!"}
http://localhost:8000/wrong [400] {"message":"Wrong?"}
http://localhost:8000/check [400] Bad status line "Invalid method encountered:\n\n  b'Host: localhost:8000\\r'\n                       ^"
```
5. Observe the console of the server. It logs a warning _Invalid HTTP request received._

### Expected behavior

Just as with the reference implementation using `requests` the failed upload attempt (2nd request) should not have an effect on the outcome of the 3rd request. The output using `aiohttp` should be the same as with `requests`:
```
http://localhost:8000/check [200] {"message":"Check!"}
http://localhost:8000/wrong [400] {"message":"Wrong?"}
http://localhost:8000/check [200] {"message":"Check!"}
```

### Logs/tracebacks

```python-traceback
$ python /home/peter/example/client_aiohttp.py 
http://localhost:8000/check [200] {"message":"Check!"}
http://localhost:8000/wrong [400] {"message":"Wrong?"}
Traceback (most recent call last):
  File "/home/peter/example/client_aiohttp.py", line 48, in <module>
    asyncio.run(main=main())
  File "/home/peter/miniconda3/envs/aiohttp/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/home/peter/miniconda3/envs/aiohttp/lib/python3.9/asyncio/base_events.py", line 647, in run_until_complete
    return future.result()
  File "/home/peter/example/client_aiohttp.py", line 42, in main
    response = await session.request(method="GET", url=check_url)
  File "/home/peter/miniconda3/envs/aiohttp/lib/python3.9/site-packages/aiohttp/client.py", line 560, in _request
    await resp.start(conn)
  File "/home/peter/miniconda3/envs/aiohttp/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 914, in start
    self._continue = None
  File "/home/peter/miniconda3/envs/aiohttp/lib/python3.9/site-packages/aiohttp/helpers.py", line 721, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
ERROR conda.cli.main_run:execute(47): `conda run python /home/peter/example/client_aiohttp.py` failed. (See above for error)
```

### Python Version

```console
$ python --version
Python 3.9.17
```

### aiohttp Version

```console
$ python -m pip show aiohttp
Name: aiohttp
Version: 3.8.5
Summary: Async http client/server framework (asyncio)
Home-page: https://github.com/aio-libs/aiohttp
Author:
Author-email:
License: Apache 2
Location: /home/peter/miniconda3/envs/aiohttp/lib/python3.9/site-packages
Requires: aiosignal, async-timeout, attrs, charset-normalizer, frozenlist, multidict, yarl
Required-by:
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.2
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /home/peter/miniconda3/envs/aiohttp/lib/python3.9/site-packages
Requires:
Required-by: aiohttp, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.7.2
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /home/peter/miniconda3/envs/aiohttp/lib/python3.9/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

### OS

Ubuntu 20.04 on WSL2 on Windows 11
```
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.6 LTS
Release:        20.04
Codename:       focal
```

### Related component

Client

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

**Comment by Dreamsorcerer** [comment]
I think there's a duplicate somewhere, possibly related to #5238 or something similar.

If you could rewrite that as a test in a PR (obviously, you'll need to use aiohttp, instead of fastapi), that'd get us a good way to making sure this is fixed.

**Comment by peterbandi** [comment]
I have updated the description to `aiohttp` only. I'll try to add a testcase later.

**Comment by webknjaz** [comment]
@peterbandi if you make it into a PR, we can merge that test without a fix (marking it with xfail). Then, whoever gets to fix it, will have something to rely on, already in Git.

**Comment by peterbandi** [comment]
@webknjaz I am not very familiar with `aiohttp` so I need some time to figure out how to write a test case where the server rejects the upload. I will try to find the time to make a PR this week. If I can't then I'll make a PR next week.

**Comment by Dreamsorcerer** [comment]
This may be a good start, there are tests in there with server handlers: https://github.com/aio-libs/aiohttp/blob/master/tests/test_client_functional.py

**Comment by peterbandi** [comment]
@Dreamsorcerer thanks. I have added a test case and made a PR for it.

**Comment by Dreamsorcerer** [comment]
Hmm, original reproducer works fine now, but xfail test is still failing...

**Comment by Dreamsorcerer** [comment]
OK, I think the original reproducer got caught by a one of the parser issues that has been fixed in recent releases.
The test sends a much larger body, which reveals another issue. Because the server doesn't read the body, the client cancels sending the full body. The client was then reusing the connection for the next request, which is unsafe and lead to the server not recognising that the next message had started.
